pkg/executor/cluster.go

```diff
@@ -15,79 +15,68 @@
 package executor
 
 import (
-	"encoding/json"
+	"context"
 
 	kdv1 "github.com/bluek8s/kubedirector/pkg/apis/kubedirector.bluedata.io/v1alpha1"
 	"github.com/bluek8s/kubedirector/pkg/shared"
-	"github.com/operator-framework/operator-sdk/pkg/sdk"
 	"k8s.io/apimachinery/pkg/types"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// statusPatchSpec is used to create PATCH operation input for modifying a
-// virtual cluster CR's status.
-type statusPatchSpec struct {
-	Op    string             `json:"op"`
-	Path  string             `json:"path"`
-	Value kdv1.ClusterStatus `json:"value"`
-}
-
-// finalizersPatchSpec is used to create PATCH operation input for modifying a
-// virtual cluster CR's finalizers list.
-type finalizersPatchSpec struct {
-	Op    string   `json:"op"`
-	Path  string   `json:"path"`
-	Value []string `json:"value"`
-}
-
 // UpdateStatus propagates status changes back to k8s. Roles or members in
 // the status that have been marked for deletion (by having certain fields
 // set to emptystring) will be removed before the writeback.
 func UpdateStatus(
 	cr *kdv1.KubeDirectorCluster,
+	client k8sclient.Client,
 ) error {
 
+	prevCr := &kdv1.KubeDirectorCluster{}
+	err := client.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name: cr.Name,
+		},
+		prevCr,
+	)
+	if err != nil {
+		shared.LogErrorf(
+			cr,
+			shared.EventReasonCluster,
+			"failed to retrieve cluster: %v",
+			err,
+		)
+		return err
+	}
+
 	// Before writing back, remove any RoleStatus where StatefulSet is
 	// emptystring, and remove any MemberStatus where Pod is emptystring.
 	compact(&(cr.Status.Roles))
 
 	// TBD: We should probably write to the status sub-resource. That's only
 	// available in 1.11 (beta feature) and later though. So for now let's
 	// just modify the status property of the CR.
-	statusPatch := []statusPatchSpec{
-		{
-			Op:    "replace",
-			Path:  "/status",
-			Value: *(cr.Status),
-		},
-	}
-	patchAttempt := func(p []statusPatchSpec) error {
-		statusPatchBytes, err := json.Marshal(statusPatch)
-		if err == nil {
-			err = sdk.Patch(cr, types.JSONPatchType, statusPatchBytes)
-		}
-		return err
-	}
-	patchErr := patchAttempt(statusPatch)
-	if patchErr != nil {
-		// If replace doesn't work, try add. (First time update.)
-		statusPatch[0].Op = "add"
-		patchErr = patchAttempt(statusPatch)
-	}
-	if patchErr != nil {
+
+	// TODO: Can we just use the cr that was passed in?
+	prevCr.Status = cr.Status
```

Comment:
More comments about PATCH stuff above.

Reply:
Until a stabile version of controller-runtime supports patch I've implemented the follow the following from Joel's recommendations:
-	Start with the given input object as the object-to-modify.
-	Apply the given modify function to the object-to-modify.
-	Try to PUT that object.
-	If that succeeds, we can return.
-	If it failed for reasons other than a resourceVersion conflict just return an error.
-	If there was a resourceVersion conflict, do a GET and modify a more recent version of the object.
-	If that fails return the error, return nil on success.
I decided to not add retry logic for the updates since no retry logic existed previously. It can easily be added later if deemed necessary. Retry logic already existed for cluster UpdateStatus and is still in place.

Reply:
Previously we didn't need retry logic in our code because we were not vulnerable to resourceVersion conflicts.

But it may be the case, at least for some of these PATCHes, that the right thing to do is fail and bail out of the reconcile handler (or at least that portion of the handler, if subsequent operations don't depend on the PATCH being done) and then let the next handler invocation try again. Basically move the retry logic out of our code and into the SDK framework. The advantage of that approach would be that we don't need to have some cap on number of retries and logic about what to do when that cap is hit.

We'll need to review the current PATCH invocations to see what will happen if we immediately bail out on error.